### DISCUSSION TITLE
Scala Pattern Matching on Java Records

### DISCUSSION THREAD

**Discussion by 3xau1o** [scala/scala3#18115] [Feature Requests]
Java 21 added a very basic [pattern matching for Java records](https://openjdk.org/jeps/440), following @smarter's [feedback](https://discord.com/channels/632150470000902164/632628489719382036/1108445810233974836)
there is chance Scala can take advantage of them.

![image](https://github.com/lampepfl/dotty/assets/19735362/3fc831fe-f9fc-4959-aa2c-8527fbf202f5)

**Comment by He-Pin** [comment]
The Java 21 will be the be the next Java 6 and 8, so +1 for this.

**Reply by He-Pin** [reply]
+1 for this, because I'm defining ADT in Java 21 and the logic in Scala, it would be great if Java record works out of the box.

**Comment by SethTisue** [comment]
I've converted this to an issue: https://github.com/scala/scala3/issues/20561  as it seems almost certain we'll tackle it eventually, perhaps even relatively soon (?), and having the issue will help us track it
